deep_speech_2/ctc_beam_search_decoder.py

```diff
@@ -0,0 +1,192 @@
+## This is a prototype of ctc beam search decoder
+
+import copy
+import random
+import numpy as np
+
+# vocab = blank + space + English characters
+#vocab = ['-', ' '] + [chr(i) for i in range(97, 123)]
+
+vocab = ['-', '_', 'a']
+
+
+def ids_list2str(ids_list):
+    ids_str = [str(elem) for elem in ids_list]
+    ids_str = ' '.join(ids_str)
+    return ids_str
+
+
+def ids_id2token(ids_list):
+    ids_str = ''
+    for ids in ids_list:
+        ids_str += vocab[ids]
+    return ids_str
+
+
+def language_model(ids_list, vocabulary):
+    # lookup ptb vocabulary
+    ptb_vocab_path = "./data/ptb_vocab.txt"
+    sentence = ''.join([vocabulary[ids] for ids in ids_list])
+    words = sentence.split(' ')
+    last_word = words[-1]
+    with open(ptb_vocab_path, 'r') as ptb_vocab:
+        f = ptb_vocab.readline()
+        while f:
+            if f == last_word:
+                return 1.0
+            f = ptb_vocab.readline()
+    return 0.0
+
+
+def ctc_beam_search_decoder(input_probs_matrix,
+                            beam_size,
+                            vocabulary,
+                            max_time_steps=None,
+                            lang_model=language_model,
+                            alpha=1.0,
+                            beta=1.0,
+                            blank_id=0,
+                            space_id=1,
+                            num_results_per_sample=None):
+    '''
+    Beam search decoder for CTC-trained network,  adapted from Algorithm 1 
+    in https://arxiv.org/abs/1408.2873.
+
+    :param input_probs_matrix: probs matrix for input sequence, row major
+    :type input_probs_matrix: 2D matrix. 
+    :param beam_size: width for beam search
+    :type beam_size: int
+    :max_time_steps: maximum steps' number for input sequence, 
+                     <=len(input_probs_matrix)
+    :type max_time_steps: int
+    :lang_model: language model for scoring
+    :type lang_model: function
+    :param alpha: parameter associated with language model.
+    :type alpha: float 
+    :param beta: parameter associated with word count
+    :type beta: float
+    :param blank_id: id of blank, default 0.
+    :type blank_id: int
+    :param space_id: id of space, default 1.
+    :type space_id: int
+    :param num_result_per_sample: the number of output decoding results 
+                                 per given sample, <=beam_size.
+    :type num_result_per_sample: int
+    '''
+
+    # function to convert ids in string to list
+    def ids_str2list(ids_str):
+        ids_str = ids_str.split(' ')
+        ids_list = [int(elem) for elem in ids_str]
+        return ids_list
+
+    # counting words in a character list
+    def word_count(ids_list):
```

Comment:
Word count scorer should be a part of the external_scoring_function handler input. Please move `word_count` to external_scoring_function, inside of which  a LM evaluator and a word-count evaluator is called.